src/core/utils/hex.ts

```diff
@@ -21,14 +22,39 @@ export const isHexStringIgnorePrefix = (value: string): boolean => {
  * @param value The starting string.
  * @return The prefixed string.
  */
-export const addHexPrefix = (value: string): string =>
-  startsWith(value, '0x') ? value : `0x${value}`;
+export const addHexPrefix = (value: string): `0x${string}` =>
+  startsWith(value, '0x') ? (value as `0x${string}`) : `0x${value}`;
 
-export const convertStringToHex = (stringToConvert: string): string =>
-  new BigNumber(stringToConvert).toString(16);
+export const convertStringToHex = (
+  stringToConvert: BigNumberish | bigint | EthersBigNumber | Bytes,
+): string => {
+  if (typeof stringToConvert === 'bigint') {
+    return stringToConvert.toString(16);
+  }
+  if (EthersBigNumber.isBigNumber(stringToConvert)) {
+    return stringToConvert.toHexString();
+  }
+  if (isBytes(stringToConvert)) {
+    return EthersBigNumber.from(stringToConvert).toHexString();
+  }
+  return new BigNumber(stringToConvert).toString(16);
+};
+
+export const toHex = (
+  stringToConvert: BigNumberish | bigint | EthersBigNumber | Bytes,
+): `0x${string}` => addHexPrefix(convertStringToHex(stringToConvert));
 
-export const toHex = (stringToConvert: string): string =>
-  addHexPrefix(convertStringToHex(stringToConvert));
+export const toHexOrUndefined = <
+  T extends BigNumberish | bigint | EthersBigNumber | Bytes | undefined,
+>(
+  stringToConvert: T,
+): T extends undefined ? undefined : `0x${string}` => {
+  if (stringToConvert === undefined)
+    return undefined as T extends undefined ? undefined : `0x${string}`;
+  return toHex(stringToConvert) as T extends undefined
+    ? undefined
+    : `0x${string}`;
+};
 
 export const toHexNoLeadingZeros = (value: string): string =>
   toHex(value).replace(/^0x0*/, '0x');
```

Comment:
Maybe a different project or refactor PR to build on, but it looks like most of these utils are implemented in ox or viem and probably don't need to exist